framework/audio/engine/internal/audioengine.cpp

```diff
@@ -0,0 +1,197 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2020 MuseScore BVBA and others
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+
+#include "audioengine.h"
+
+#include <soloud.h>
+
+#include "log.h"
+#include "ptrutils.h"
+
+#include "modularity/ioc.h"
+#include "iaudiodriver.h"
+
+using namespace mu::audio::engine;
+
+constexpr int BUF_SIZE{ 1024 };
+
+struct AudioEngine::SL {
+    SoLoud::Soloud engine;
+};
+
+AudioEngine::AudioEngine()
+{
+    m_sl = std::unique_ptr<SL>(new SL);
+}
+
+AudioEngine::~AudioEngine()
+{
+}
+
+bool AudioEngine::isInited() const
+{
+    std::shared_ptr<IAudioDriver> drv = driver();
+    if (!drv) {
+        return false;
+    }
+
+    if (!drv->isOpened()) {
+        return false;
+    }
+
+    if (!m_inited) {
+        return false;
+    }
+
+    return true;
+}
+
+bool AudioEngine::init()
+{
+    if (isInited()) {
+        return true;
+    }
+
+    auto res = m_sl->engine.init(SoLoud::Soloud::CLIP_ROUNDOFF,
+                                 SoLoud::Soloud::MUAUDIO,
+                                 SoLoud::Soloud::AUTO,
+                                 BUF_SIZE,
+                                 2);
+
+    if (res == SoLoud::SO_NO_ERROR) {
+        LOGI() << "success inited audio engine";
+        m_inited = true;
+    } else {
+        LOGE() << "failed inited audio engine, err: " << res;
+        m_inited = false;
+    }
+
+    return m_inited;
+}
+
+void AudioEngine::deinit()
+{
+    m_sl->engine.deinit();
+    m_inited = false;
+}
+
+float AudioEngine::sampleRate() const
+{
+    return m_sl->engine.getBackendSamplerate();
+}
+
+IAudioEngine::handle AudioEngine::play(IAudioSource* s, float volume, float pan, bool paused)
```

Comment:
What does paused affect?

Reply:
This means that the sound source will be initialized and added to the playlist, but will be paused. This is useful when you need to start playing several sources at the same time, everyone is added to the play in pause mode, and then you do unpause all.